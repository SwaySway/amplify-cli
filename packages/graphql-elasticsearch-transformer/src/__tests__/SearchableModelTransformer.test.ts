import {
  parse,
  ObjectTypeDefinitionNode,
  Kind,
  DefinitionNode,
  DocumentNode,
  FieldDefinitionNode,
  InputObjectTypeDefinitionNode,
} from 'graphql';
import { GraphQLTransform, InvalidDirectiveError } from 'graphql-transformer-core';
import { SearchableModelTransformer } from '../SearchableModelTransformer';
import { DynamoDBModelTransformer } from 'graphql-dynamodb-transformer';

test('schema without @model', () => {
  const invalidSchema = `
  type Record @searchable {
    id: ID!
    createdAt: AWSDate
    updatedAt: AWSDateTime
    info: String
    precision: Float
    age: Int
    active: Boolean
    email: AWSEmail
    json: AWSJSON
    url: AWSURL
    phone: AWSPhone
    ip: AWSIPAddress
    time: AWSTime
  }`;

  const transformer = new GraphQLTransform({
    transformers: [new SearchableModelTransformer()],
  });
  expect(() => transformer.transform(invalidSchema)).toThrowError(InvalidDirectiveError);
});

test('filter input types have the correct output', () => {
  const validSchema = `
  type Record @model @searchable {
    id: ID!
    createdAt: AWSDate
    updatedAt: AWSDateTime
    info: String
    precision: Float
    age: Int
    active: Boolean
    email: AWSEmail
    json: AWSJSON
    url: AWSURL
    phone: AWSPhone
    ip: AWSIPAddress
    time: AWSTime
  }`;

  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new SearchableModelTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toContain('input SearchableDateFilterInput');
  expect(out.schema).toMatchSnapshot();
});
test('Test SearchableModelTransformer with query overrides', () => {
  const validSchema = `type Post @model @searchable(queries: { search: "customSearchPost" }) {
        id: ID!
        title: String!
        createdAt: String
        updatedAt: String
    }`;

  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new SearchableModelTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toBeDefined();
  const parsed = parse(out.schema);
  const queryType = getObjectType(parsed, 'Query');
  expect(queryType).toBeDefined();
  expectFields(queryType, ['customSearchPost']);
});

test('Test SearchableModelTransformer with only create mutations', () => {
  const validSchema = `type Post @model(mutations: { create: "customCreatePost" }) @searchable {
        id: ID!
        title: String!
        createdAt: String
        updatedAt: String
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new SearchableModelTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  const parsed = parse(out.schema);
  const mutationType = getObjectType(parsed, 'Mutation');
  expect(mutationType).toBeDefined();
  expectFields(mutationType, ['customCreatePost']);
  doNotExpectFields(mutationType, ['updatePost']);
});

test('Test SearchableModelTransformer with multiple model searchable directives', () => {
  const validSchema = `
    type Post @model @searchable {
        id: ID!
        title: String!
        createdAt: String
        updatedAt: String
    }

    type User @model @searchable {
        id: ID!
        name: String!
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new SearchableModelTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toBeDefined();
  const parsed = parse(out.schema);
  const queryType = getObjectType(parsed, 'Query');
  expect(queryType).toBeDefined();
  expectFields(queryType, ['searchPosts']);
  expectFields(queryType, ['searchUsers']);

  const stringInputType = getInputType(parsed, 'SearchableStringFilterInput');
  expect(stringInputType).toBeDefined();
  const booleanInputType = getInputType(parsed, 'SearchableBooleanFilterInput');
  expect(booleanInputType).toBeDefined();
  const intInputType = getInputType(parsed, 'SearchableIntFilterInput');
  expect(intInputType).toBeDefined();
  const floatInputType = getInputType(parsed, 'SearchableFloatFilterInput');
  expect(floatInputType).toBeDefined();
  const dateInputType = getInputType(parsed, 'SearchableDateFilterInput');
  expect(dateInputType).toBeDefined();
  const idInputType = getInputType(parsed, 'SearchableIDFilterInput');
  expect(idInputType).toBeDefined();
  const postInputType = getInputType(parsed, 'SearchablePostFilterInput');
  expect(postInputType).toBeDefined();
  const userInputType = getInputType(parsed, 'SearchableUserFilterInput');
  expect(userInputType).toBeDefined();

  expect(verifyInputCount(parsed, 'SearchableStringFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchableBooleanFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchableIntFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchableFloatFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'ModelIDFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchableDateFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchablePostFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchableUserFilterInput', 1)).toBeTruthy;
});

test('Test SearchableModelTransformer with sort fields', () => {
  const validSchema = `
    type Post @model @searchable {
        id: ID!
        title: String!
        createdAt: String
        updatedAt: String
    }`;

  const transformer = new GraphQLTransform({
    transformers: [new DynamoDBModelTransformer(), new SearchableModelTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toBeDefined();
  const parsed = parse(out.schema);
  const queryType = getObjectType(parsed, 'Query');
  expect(queryType).toBeDefined();
  expectFields(queryType, ['searchPosts']);

  const stringInputType = getInputType(parsed, 'SearchableStringFilterInput');
  expect(stringInputType).toBeDefined();
  const booleanInputType = getInputType(parsed, 'SearchableBooleanFilterInput');
  expect(booleanInputType).toBeDefined();
  const intInputType = getInputType(parsed, 'SearchableIntFilterInput');
  expect(intInputType).toBeDefined();
  const floatInputType = getInputType(parsed, 'SearchableFloatFilterInput');
  expect(floatInputType).toBeDefined();
  const dateInputType = getInputType(parsed, 'SearchableDateFilterInput');
  expect(dateInputType).toBeDefined();
  const idInputType = getInputType(parsed, 'SearchableIDFilterInput');
  expect(idInputType).toBeDefined();
  const postInputType = getInputType(parsed, 'SearchablePostFilterInput');
  expect(postInputType).toBeDefined();
  const sortInputType = getInputType(parsed, 'SearchablePostSortInput');
  expect(sortInputType).toBeDefined();

  expect(verifyInputCount(parsed, 'SearchableStringFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchableBooleanFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchableIntFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchableFloatFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchableDateFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchablePostFilterInput', 1)).toBeTruthy;
  expect(verifyInputCount(parsed, 'SearchablePostSortInput', 1)).toBeTruthy;
});

function expectFields(type: ObjectTypeDefinitionNode, fields: string[]) {
  for (const fieldName of fields) {
    const foundField = type.fields.find((f: FieldDefinitionNode) => f.name.value === fieldName);
    expect(foundField).toBeDefined();
  }
}

function doNotExpectFields(type: ObjectTypeDefinitionNode, fields: string[]) {
  for (const fieldName of fields) {
    expect(type.fields.find((f: FieldDefinitionNode) => f.name.value === fieldName)).toBeUndefined();
  }
}

function getObjectType(doc: DocumentNode, type: string): ObjectTypeDefinitionNode | undefined {
  return doc.definitions.find((def: DefinitionNode) => def.kind === Kind.OBJECT_TYPE_DEFINITION && def.name.value === type) as
    | ObjectTypeDefinitionNode
    | undefined;
}

function getInputType(doc: DocumentNode, type: string): InputObjectTypeDefinitionNode | undefined {
  return doc.definitions.find((def: DefinitionNode) => def.kind === Kind.INPUT_OBJECT_TYPE_DEFINITION && def.name.value === type) as
    | InputObjectTypeDefinitionNode
    | undefined;
}

function verifyInputCount(doc: DocumentNode, type: string, count: number): boolean {
  return doc.definitions.filter(def => def.kind === Kind.INPUT_OBJECT_TYPE_DEFINITION && def.name.value === type).length == count;
}
