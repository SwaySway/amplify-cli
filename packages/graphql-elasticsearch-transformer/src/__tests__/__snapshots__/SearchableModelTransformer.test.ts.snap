// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter input types have the correct output 1`] = `
"type Record {
  id: ID!
  createdAt: AWSDate
  updatedAt: AWSDateTime
  info: String
  precision: Float
  age: Int
  active: Boolean
  email: AWSEmail
  json: AWSJSON
  url: AWSURL
  phone: AWSPhone
  ip: AWSIPAddress
  time: AWSTime
}

enum ModelSortDirection {
  ASC
  DESC
}

type ModelRecordConnection {
  items: [Record]
  nextToken: String
}

input ModelStringFilterInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
}

input ModelIDFilterInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
}

input ModelIntFilterInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

input ModelFloatFilterInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
}

input ModelBooleanFilterInput {
  ne: Boolean
  eq: Boolean
}

input ModelRecordFilterInput {
  id: ModelIDFilterInput
  createdAt: ModelStringFilterInput
  updatedAt: ModelStringFilterInput
  info: ModelStringFilterInput
  precision: ModelFloatFilterInput
  age: ModelIntFilterInput
  active: ModelBooleanFilterInput
  email: ModelStringFilterInput
  json: ModelStringFilterInput
  url: ModelStringFilterInput
  phone: ModelStringFilterInput
  ip: ModelStringFilterInput
  time: ModelStringFilterInput
  and: [ModelRecordFilterInput]
  or: [ModelRecordFilterInput]
  not: ModelRecordFilterInput
}

type Query {
  getRecord(id: ID!): Record
  listRecords(filter: ModelRecordFilterInput, limit: Int, nextToken: String): ModelRecordConnection
  searchRecords(filter: SearchableRecordFilterInput, sort: SearchableRecordSortInput, limit: Int, nextToken: String): SearchableRecordConnection
}

input CreateRecordInput {
  id: ID
  createdAt: AWSDate
  updatedAt: AWSDateTime
  info: String
  precision: Float
  age: Int
  active: Boolean
  email: AWSEmail
  json: AWSJSON
  url: AWSURL
  phone: AWSPhone
  ip: AWSIPAddress
  time: AWSTime
}

input UpdateRecordInput {
  id: ID!
  createdAt: AWSDate
  updatedAt: AWSDateTime
  info: String
  precision: Float
  age: Int
  active: Boolean
  email: AWSEmail
  json: AWSJSON
  url: AWSURL
  phone: AWSPhone
  ip: AWSIPAddress
  time: AWSTime
}

input DeleteRecordInput {
  id: ID
}

type Mutation {
  createRecord(input: CreateRecordInput!): Record
  updateRecord(input: UpdateRecordInput!): Record
  deleteRecord(input: DeleteRecordInput!): Record
}

type Subscription {
  onCreateRecord: Record @aws_subscribe(mutations: [\\"createRecord\\"])
  onUpdateRecord: Record @aws_subscribe(mutations: [\\"updateRecord\\"])
  onDeleteRecord: Record @aws_subscribe(mutations: [\\"deleteRecord\\"])
}

input SearchableStringFilterInput {
  ne: String
  gt: String
  lt: String
  gte: String
  lte: String
  eq: String
  match: String
  matchPhrase: String
  matchPhrasePrefix: String
  multiMatch: String
  exists: Boolean
  wildcard: String
  regexp: String
}

input SearchableIDFilterInput {
  ne: ID
  gt: ID
  lt: ID
  gte: ID
  lte: ID
  eq: ID
  match: ID
  matchPhrase: ID
  matchPhrasePrefix: ID
  multiMatch: ID
  exists: Boolean
  wildcard: ID
  regexp: ID
}

input SearchableIntFilterInput {
  ne: Int
  gt: Int
  lt: Int
  gte: Int
  lte: Int
  eq: Int
  range: [Int]
}

input SearchableFloatFilterInput {
  ne: Float
  gt: Float
  lt: Float
  gte: Float
  lte: Float
  eq: Float
  range: [Float]
}

input SearchableBooleanFilterInput {
  eq: Boolean
  ne: Boolean
}

input SearchableDateFilterInput {
  ne: String
  gt: String
  lt: String
  gte: String
  lte: String
  eq: String
  exists: Boolean
  range: [String]
}

input SearchableRecordFilterInput {
  id: SearchableIDFilterInput
  createdAt: SearchableDateFilterInput
  updatedAt: SearchableDateFilterInput
  info: SearchableStringFilterInput
  precision: SearchableFloatFilterInput
  age: SearchableIntFilterInput
  active: SearchableBooleanFilterInput
  email: SearchableStringFilterInput
  json: SearchableStringFilterInput
  url: SearchableStringFilterInput
  phone: SearchableStringFilterInput
  ip: SearchableStringFilterInput
  time: SearchableStringFilterInput
  and: [SearchableRecordFilterInput]
  or: [SearchableRecordFilterInput]
  not: SearchableRecordFilterInput
}

enum SearchableSortDirection {
  asc
  desc
}

enum SearchableRecordSortableFields {
  id
  createdAt
  updatedAt
  info
  precision
  age
  active
  email
  json
  url
  phone
  ip
  time
}

input SearchableRecordSortInput {
  field: SearchableRecordSortableFields
  direction: SearchableSortDirection
}

type SearchableRecordConnection {
  items: [Record]
  nextToken: String
  total: Int
}
"
`;
